The changes look good to me, but Conformance E2E is consistently failing. It seems that there's something wrong  with kube-apiserver.

https://github.com/cilium/cilium/actions/runs/6932811318/job/18884158732
```
Error: Unable to install Cilium: Get "https://127.0.0.1:35039/api/v1/namespaces/kube-system/services/hubble-peer": http2: client connection lost
Error: Process completed with exit code 1.
```

--- a/bpf/bpf_host.c
+++ b/bpf/bpf_host.c
@@ -173,7 +173,9 @@ handle_ipv6(struct __ctx_buff *ctx, __u32 secctx __maybe_unused,
 #ifdef ENABLE_NODEPORT
 	if (!from_host) {
 		if (!ctx_skip_nodeport(ctx)) {
-			ret = nodeport_lb6(ctx, ip6, secctx, ext_err);
+			bool is_dsr = false;
+
+			ret = nodeport_lb6(ctx, ip6, secctx, ext_err, &is_dsr);
 			/* nodeport_lb6() returns with TC_ACT_REDIRECT for
 			 * traffic to L7 LB. Policy enforcement needs to take
 			 * place after L7 LB has processed the packet, so we
@@ -564,7 +566,9 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx __maybe_unused,
 #ifdef ENABLE_NODEPORT
 	if (!from_host) {
 		if (!ctx_skip_nodeport(ctx)) {
-			int ret = nodeport_lb4(ctx, ip4, ETH_HLEN, secctx, ext_err);
+			bool __maybe_unused is_dsr = false;
+
+			int ret = nodeport_lb4(ctx, ip4, ETH_HLEN, secctx, ext_err, &is_dsr);
 
 			if (ret == NAT_46X64_RECIRC) {
 				ctx_store_meta(ctx, CB_SRC_LABEL, secctx);

--- a/bpf/bpf_overlay.c
+++ b/bpf/bpf_overlay.c
@@ -50,13 +50,14 @@ static __always_inline int handle_ipv6(struct __ctx_buff *ctx,
 	struct ipv6hdr *ip6;
 	struct endpoint_info *ep;
 	bool decrypted;
+	bool __maybe_unused is_dsr = false;
 
 	/* verifier workaround (dereference of modified ctx ptr) */
 	if (!revalidate_data_pull(ctx, &data, &data_end, &ip6))
 		return DROP_INVALID;
 #ifdef ENABLE_NODEPORT
 	if (!ctx_skip_nodeport(ctx)) {
-		ret = nodeport_lb6(ctx, ip6, *identity, ext_err);
+		ret = nodeport_lb6(ctx, ip6, *identity, ext_err, &is_dsr);
 		/* nodeport_lb6() returns with TC_ACT_REDIRECT for
 		 * traffic to L7 LB. Policy enforcement needs to take
 		 * place after L7 LB has processed the packet, so we
@@ -86,9 +87,13 @@ static __always_inline int handle_ipv6(struct __ctx_buff *ctx,
 	} else {
 		/* Maybe overwrite the REMOTE_NODE_ID with
 		 * KUBE_APISERVER_NODE_ID to support upgrade. After v1.12,
-		 * this should be removed.
+		 * identity_is_remote_node() should be removed.
+		 *
+		 * A packet that has DSR info and comes from `world` may have specific identity when
+		 * a CNP that is using CIDR rules is applied.
 		 */
-		if (info && identity_is_remote_node(*identity))
+		if (info && (identity_is_remote_node(*identity) ||
+			     (is_dsr && identity_is_world_ipv6(*identity))))
 			*identity = info->sec_identity;
 	}
 
@@ -271,6 +276,7 @@ static __always_inline int handle_ipv4(struct __ctx_buff *ctx,
 	struct iphdr *ip4;
 	struct endpoint_info *ep;
 	bool decrypted;
+	bool __maybe_unused is_dsr = false;
 
 	/* verifier workaround (dereference of modified ctx ptr) */
 	if (!revalidate_data_pull(ctx, &data, &data_end, &ip4))
@@ -287,7 +293,7 @@ static __always_inline int handle_ipv4(struct __ctx_buff *ctx,
 
 #ifdef ENABLE_NODEPORT
 	if (!ctx_skip_nodeport(ctx)) {
-		int ret = nodeport_lb4(ctx, ip4, ETH_HLEN, *identity, ext_err);
+		int ret = nodeport_lb4(ctx, ip4, ETH_HLEN, *identity, ext_err, &is_dsr);
 		/* nodeport_lb4() returns with TC_ACT_REDIRECT for
 		 * traffic to L7 LB. Policy enforcement needs to take
 		 * place after L7 LB has processed the packet, so we
@@ -353,7 +359,8 @@ skip_vtep:
 		}
 #endif
 		/* See comment at equivalent code in handle_ipv6() */
-		if (info && identity_is_remote_node(*identity))
+		if (info && (identity_is_remote_node(*identity) ||
+			     (is_dsr && identity_is_world_ipv4(*identity))))
 			*identity = info->sec_identity;
 	}
 

--- a/bpf/bpf_xdp.c
+++ b/bpf/bpf_xdp.c
@@ -104,6 +104,7 @@ int tail_lb_ipv4(struct __ctx_buff *ctx)
 		int l3_off = ETH_HLEN;
 		void *data, *data_end;
 		struct iphdr *ip4;
+		bool __maybe_unused is_dsr = false;
 
 		if (!revalidate_data(ctx, &data, &data_end, &ip4)) {
 			ret = DROP_INVALID;
@@ -184,7 +185,7 @@ int tail_lb_ipv4(struct __ctx_buff *ctx)
 no_encap:
 #endif /* ENABLE_DSR && !ENABLE_DSR_HYBRID && DSR_ENCAP_MODE == DSR_ENCAP_GENEVE */
 
-		ret = nodeport_lb4(ctx, ip4, l3_off, 0, &ext_err);
+		ret = nodeport_lb4(ctx, ip4, l3_off, 0, &ext_err, &is_dsr);
 		if (ret == NAT_46X64_RECIRC) {
 			ep_tail_call(ctx, CILIUM_CALL_IPV6_FROM_NETDEV);
 			return send_drop_notify_error(ctx, 0, DROP_MISSED_TAIL_CALL,
@@ -257,13 +258,14 @@ int tail_lb_ipv6(struct __ctx_buff *ctx)
 	if (!ctx_skip_nodeport(ctx)) {
 		void *data, *data_end;
 		struct ipv6hdr *ip6;
+		bool is_dsr = false;
 
 		if (!revalidate_data(ctx, &data, &data_end, &ip6)) {
 			ret = DROP_INVALID;
 			goto drop_err;
 		}
 
-		ret = nodeport_lb6(ctx, ip6, 0, &ext_err);
+		ret = nodeport_lb6(ctx, ip6, 0, &ext_err, &is_dsr);
 		if (IS_ERR(ret))
 			goto drop_err;
 	}

--- a/bpf/lib/nodeport.h
+++ b/bpf/lib/nodeport.h
@@ -1270,7 +1270,8 @@ drop_err:
 static __always_inline int nodeport_lb6(struct __ctx_buff *ctx,
 					struct ipv6hdr *ip6,
 					__u32 src_sec_identity,
-					__s8 *ext_err)
+					__s8 *ext_err,
+					bool __maybe_unused *dsr)
 {
 	bool is_svc_proto __maybe_unused = true;
 	int ret, l3_off = ETH_HLEN, l4_off;
@@ -1355,15 +1356,13 @@ skip_service_lookup:
 #if (defined(IS_BPF_OVERLAY) && DSR_ENCAP_MODE == DSR_ENCAP_GENEVE) || \
 	(!defined(IS_BPF_OVERLAY) && DSR_ENCAP_MODE != DSR_ENCAP_GENEVE)
 		if (is_svc_proto && nodeport_uses_dsr6(&tuple)) {
-			bool dsr = false;
-
 			ret = nodeport_extract_dsr_v6(ctx, ip6, &tuple, l4_off,
 						      &key.address,
-						      &key.dport, &dsr);
+						      &key.dport, dsr);
 			if (IS_ERR(ret))
 				return ret;
 
-			if (dsr)
+			if (*dsr)
 				return nodeport_dsr_ingress_ipv6(ctx, &tuple, l4_off,
 								 &key.address, key.dport,
 								 ext_err);
@@ -2747,7 +2746,8 @@ static __always_inline int nodeport_lb4(struct __ctx_buff *ctx,
 					struct iphdr *ip4,
 					int l3_off,
 					__u32 src_sec_identity,
-					__s8 *ext_err)
+					__s8 *ext_err,
+					bool __maybe_unused *dsr)
 {
 	bool is_fragment = ipv4_is_fragment(ip4);
 	bool backend_local, has_l4_header;
@@ -2840,18 +2840,16 @@ skip_service_lookup:
 #if (defined(IS_BPF_OVERLAY) && DSR_ENCAP_MODE == DSR_ENCAP_GENEVE) || \
 	(!defined(IS_BPF_OVERLAY) && DSR_ENCAP_MODE != DSR_ENCAP_GENEVE)
 		if (is_svc_proto && nodeport_uses_dsr4(&tuple)) {
-			bool dsr = false;
-
 			/* Check if packet has embedded DSR info, or belongs to
 			 * an established DSR connection:
 			 */
 			ret = nodeport_extract_dsr_v4(ctx, ip4, &tuple,
 						      l4_off, &key.address,
-						      &key.dport, &dsr);
+						      &key.dport, dsr);
 			if (IS_ERR(ret))
 				return ret;
 
-			if (dsr)
+			if (*dsr)
 				/* Packet continues on its way to local backend: */
 				return nodeport_dsr_ingress_ipv4(ctx, &tuple, ip4,
 								 has_l4_header, l4_off,

